### ISSUE TITLE
Don't crash with NPE on absolute file input paths. [phab:D950]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1274]
_Created by **benjyw** at 2016-07-08 02:55:07_: 

Calling Path.getFileName() on "/" returns null, which triggered
a NPE in CompilationUnitFileTree.  This commit adds detection for
that case.

It adds a test for this case which breaks before the fix and passes
after it.

**Comment by kythe** [comment]
_**schroederc** wrote at 2016-07-08 21:56:44_:

In what instance are you seeing absolute file paths in CompilationUnits?  Each path should be corpus-root relative based on the `KYTHE_ROOT_DIRECTORY` given to the extractor.

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-07-09 14:59:38_:

>>! In (earlier comment), schroederc wrote:
> In what instance are you seeing absolute file paths in CompilationUnits?  Each path should be corpus-root relative based on the `KYTHE_ROOT_DIRECTORY` given to the extractor.

I'm seeing it for loose .class files that were on the original compilation's classpath. These are outside the KYTHE_ROOT_DIRECTORY (they are an upstream dependency that happens to have been expanded on disk).  Unexpanded .jar files in similar situations don't have this problem because the !jar! prefix makes the path relative (from the point of view of path segment manipulation).

Also, note this comment: https://github.com/benjyw/kythe/blob/master/kythe/java/com/google/devtools/kythe/extractors/java/JavaCompilationUnitExtractor.java#L469

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-11 20:52:44_:

I think making this change will be cleaner than special-casing the `putDigest` method.

```
while ((curPath = curPath.getParent()) != null && curPath.getFileName() != null) {
```

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-11 20:52:44_:

`s/2014/2016/`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-11 20:52:44_:

This is fine:
```
/** Tests {link CompilationUnitFileTree} */
```

**Comment by kythe** [comment]
_**schroederc** wrote at 2016-07-11 20:52:44_:

Please run the `.java` files through https://github.com/google/google-java-format

**Comment by kythe** [comment]
_**schroederc** required changes at 2016-07-11 20:52:51_

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-07-11 21:42:56_:

Fixed. PTAL.

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-07-11 21:44:24_:

Address code review comments.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-11 21:49:12_:

Please revert this line.

**Comment by kythe** [comment]
_**benjyw** commented inline (context lost during migration) at 2016-07-11 21:57:19_:

Done, but may I ask why? If parent==null and path.isAbsolute() then path.getFileName() will return null and we'll get a NPE on the next line anyway.

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-07-11 21:59:33_:

Revert one line.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-12 16:37:38_:

You're right.  You can change it back and add a comment, if you want.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2016-07-12 16:37:38_

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-07-12 21:21:18_:

Add a comment and an assertion that path != '/' in putDigest.

**Comment by kythe** [comment]
_**benjyw** commented inline (context lost during migration) at 2016-07-12 21:23:04_:

Done, added a comment and also throw an IllegalArgumentException if '/' is passed in, so that the error is more explicit than the opaque NPE.   It's generally safer to assert your invariants instead of assuming them...

**Comment by kythe** [comment]
_**schroederc** closed via commit 02cac1c5df6619f92e95af1a6dd33938e4be6622 at 2016-07-14 19:35:23_
